bzip2.rs

```diff
@@ -363,460 +363,224 @@ unsafe fn compressStream(stream: *mut FILE, zStream: *mut FILE) {
     }
     ioError();
 }
-unsafe fn uncompressStream(zStream: *mut FILE, stream: *mut FILE) -> Bool {
-    let mut current_block: u64;
-    let mut bzf: *mut libc::c_void;
+unsafe fn uncompressStream(zStream: *mut FILE, stream: *mut FILE) -> bool {
+    let mut bzf = std::ptr::null_mut();
     let mut bzerr: i32 = 0;
     let mut bzerr_dummy: i32 = 0;
     let mut ret: i32;
     let mut nread: i32;
-    let mut i: i32;
     let mut obuf: [u8; 5000] = [0; 5000];
     let mut unused: [u8; 5000] = [0; 5000];
     let mut unusedTmpV: *mut libc::c_void = std::ptr::null_mut::<libc::c_void>();
+
     let mut nUnused: libc::c_int = 0;
     let mut streamNo: libc::c_int = 0;
-    if ferror(stream) == 0 && ferror(zStream) == 0 {
-        's_37: loop {
-            bzf = BZ2_bzReadOpen(
-                &mut bzerr,
-                zStream,
-                verbosity,
-                smallMode as libc::c_int,
-                unused.as_mut_ptr() as *mut libc::c_void,
-                nUnused,
-            );
-            if bzf.is_null() || bzerr != 0 as libc::c_int {
-                current_block = 12043696510059011606;
-                break;
-            }
-            streamNo += 1;
-            while bzerr == 0 as libc::c_int {
-                nread = BZ2_bzRead(
+
+    enum State {
+        Standard,
+        CloseOk,
+        TryCat,
+        ErrHandler,
+    }
+
+    let mut state = State::Standard;
+
+    // TODO: set the file into binary mode?
+
+    if ferror(stream) != 0 || ferror(zStream) != 0 {
+        // diverges
+        ioError()
+    }
+
+    'outer: loop {
+        match state {
+            State::Standard => loop {
```

Comment:
Would it be possible to structure the code something like:

```rust
'outer: {
    loop {
        // ...
        if should_goto_trycat {
            state = TryCat;
            break 'outer;
        }
        if should_goto_errhandler {
            state = ErrHandler;
            break 'outer;
        }
    }
}

'state: loop {
    match state {
        // ...
    }
}
```

given that the `Standard` state is never re-entered. Or does that make the code less clear?

Reply:
for now I went with a correct transformation. There is a bunch of stuff in there that still needs further cleanup (e.g. casts) and I suspect we can play around with this control flow a bit too.